#### Aquí encontrarás todo lo que necesitas saber: https://www.kaggle.com/t/ab8726f0cfc84544abbae69a6be88071

## Librerias

In [2]:
import pandas as pd

import urllib.request
from PIL import Image

## Data

In [88]:
### Para que funcione necesitas bajarte los archivos de datos de Kaggle 
train = pd.read_csv("data/train2.csv", index_col=0)


## Data exploration

In [89]:
train.info()

<class 'pandas.core.frame.DataFrame'>
Index: 912 entries, 1104 to 1173
Data columns (total 17 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   Company      912 non-null    int64  
 1   TypeName     912 non-null    int64  
 2   Inches       912 non-null    float64
 3   Ram          912 non-null    int64  
 4   OpSys        912 non-null    int64  
 5   Weight       912 non-null    float64
 6   Price_euros  912 non-null    float64
 7   Memory1_1    912 non-null    int64  
 8   Memory1_2    912 non-null    int64  
 9   Memory2_1    912 non-null    int64  
 10  Memory2_2    912 non-null    int64  
 11  Resolution   912 non-null    int64  
 12  Gpu1         912 non-null    int64  
 13  Gpu2         912 non-null    int64  
 14  Cpu1         912 non-null    int64  
 15  Cpu2         912 non-null    int64  
 16  Cpu3         912 non-null    int64  
dtypes: float64(3), int64(14)
memory usage: 128.2 KB


In [90]:
train.head()

Company  TypeName  Inches  Ram  OpSys  Weight  Price_euros   
laptop_ID                                                               
1104             1         5    15.6    4      6    2.40        387.0  \
114              2         4    13.3    8      6    1.23       1379.0   
172              1         5    17.3    8      6    3.00        854.0   
918              3         3    12.5    8      6    1.26       1483.0   
447              4         5    17.3    8      6    2.80        519.0   

           Memory1_1  Memory1_2  Memory2_1  Memory2_2  Resolution  Gpu1  Gpu2   
laptop_ID                                                                       
1104             500          1          0          0        1049     1     1  \
114              256          2          0          0        5760     2     2   
172              256          2          0          0        2073     3     6   
918              512          2          0          0        2073     2     3   
447             1000          1          0          0        8294     1     1   

           Cpu1  Cpu2  Cpu3  
laptop_ID                    
1104          1     2     2  
114           2     9     9  
172           2     9     9  
918           2     9     9  
447           1     1     1

## Data processing

-----------------------------------------------------------------------------------------------------------------

## Pasos que si o si debéis realizar para poder participar en la competición:

### 1. Definir X e y

In [91]:
X = train.drop(['Price_euros'], axis=1)
y = train['Price_euros'].copy()
X.shape

(912, 16)

In [92]:
y.shape

(912,)

### 2. Dividir X_train, X_test, y_train, y_test

In [93]:
from sklearn.model_selection import train_test_split

In [94]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.20, random_state = 42)

In [95]:
X_train

Company  TypeName  Inches  Ram  OpSys  Weight  Memory1_1   
laptop_ID                                                             
26               2         5    15.6    4      6    2.30       1000  \
776              2         2    15.6   16      6    2.72        128   
1263             5         2    15.6   16      6    2.50        256   
182              4         5    15.6    8      6    1.90        128   
321              3         1    13.3   16      6    1.29       1000   
...            ...       ...     ...  ...    ...     ...        ...   
394              2         6    15.6    8      6    2.00        256   
441              4         5    15.6   16      6    2.50        512   
955              5         5    17.3    4      6    2.80       1000   
210              2         5    15.6    6      6    2.30       1000   
673             10         5    15.6    4      6    2.10        128   

           Memory1_2  Memory2_1  Memory2_2  Resolution  Gpu1  Gpu2  Cpu1   
laptop_ID                                                                  
26                 1          0          0        1049     2     3     2  \
776                2       1000          1        8294     3     6     2   
1263               2       1000          1        2073     3     6     2   
182                2       1000          1        2073     1     8     2   
321                2          0          0        8294     2     4     2   
...              ...        ...        ...         ...   ...   ...   ...   
394                2          0          0        2073     3     9     2   
441                2          0          0        2073     3     9     2   
955                1          0          0        8294     3     6     2   
210                1          0          0        1049     2     3     2   
673                2          0          0        1049     2     3     2   

           Cpu2  Cpu3  
laptop_ID              
26            9     9  
776           9     9  
1263          9     9  
182           9     9  
321           9     9  
...         ...   ...  
394           9     9  
441           9     9  
955           7     7  
210           9     9  
673           9     9  

[729 rows x 16 columns]

### 3. Asignar el modelo (vacío) a una variable


In [96]:
from sklearn.linear_model import LinearRegression

model = LinearRegression()

In [97]:
model.fit(X_train, y_train)

LinearRegression()

In [98]:
prediction = model.predict(X_test)
prediction

array([1157.45783411,  376.00124824,  977.80496757, 1279.54877545,
       1946.15408832,  793.36463091,  802.64455637, 1316.63350257,
        977.80496757, 1165.45488137, 1180.88877484,  968.02549596,
        500.86459425, 1239.13621028, 1355.21144234, 1797.10683594,
       1823.04245346, 1289.73941832, 1046.84273152,  573.92130069,
        636.34832414,  153.38934873, 1126.32819237, 1171.28148886,
       1023.61625517,  417.7100286 ,  972.9925212 , 1781.85276075,
       1136.92016816,  357.54197493, 1118.13102617, 1288.18152125,
        602.98333246, 1366.144504  , 1256.11478849,  372.3160238 ,
        909.9843762 ,  827.64016469, 1168.12881913,  689.10892156,
       1111.20949734, 1445.4989396 , 1000.22005816, 1847.49947351,
       1063.72916574,  601.52899824, 1442.71422974,  637.75329268,
       1253.89719509,  505.31131973, 1895.95638239,  825.67754417,
        201.53718982, 1996.40424891,  706.50137811, 1013.08284825,
       1812.71236497, 1205.12060771, 1802.35790183, 1800.27499

### 4. Sacar métricas, valorar el modelo 

Recuerda que en la competición se va a evaluar con la métrica de MAE.

In [99]:
from sklearn.metrics import mean_squared_error
import numpy as np

print('RMSE :', np.sqrt(mean_squared_error(y_test, prediction)))

RMSE : 456.25893154524016


-----------------------------------------------------------------

## Una vez listo el modelo, toca predecir con el dataset de predicción 

Definición de **modelo que está listo**. 

Tras hacer suficientes pruebas, analizar los datos, hacer feature engineering, probar diferentes modelos con diferentes parámetros, es con este con el que observo mejores métricas y menos overfitting. ¡Cuidado con el overfitting aquí! Si vuestro modelo aprende muy bien de estos datos pero hay overfitting cuando le pasemos los datos desconocidos de `test.csv` nos arriesgamos a que digamos, no salga lo esperado.

### 1. Entrena dicho modelo con TODOS tus datos de train, esto es con `train.csv` al completo.


**CON LAS TRANSFORMACIONES QUE LE HAYAS REALIZADO A `X` INCLUÍDAS.**


Véase:
- Estandarización/Normalización
- Eliminación de Outliers
- Eliminación de columnas
- Creación de columnas nuevas
- Gestión de valores nulos
- Y un largo etcétera de técnicas que como Data Scientist hayas considerado las mejores para tu dataset.

### 2. Carga los datos de `test.csv` para predecir.

In [103]:
X_pred = pd.read_csv("data/test2.csv", index_col=0)
X_pred.head()

Company  TypeName  Inches  Ram  OpSys  Weight  Memory1_1   
laptop_ID                                                             
750             10         5    14.0    8      6    1.25        256  \
726             15         5    14.0    4      6    1.40         32   
633             15         5    14.0    2      6    1.40         32   
363              2         2    15.6    8      6    2.65       1000   
319              5         1    11.6    2      6    1.10         32   

           Memory1_2  Memory2_1  Memory2_2  Resolution  Gpu1  Gpu2  Cpu1   
laptop_ID                                                                  
750                2          0          0        2073     2     3     2  \
726                1          0          0        2304     2     3     2   
633                4          0          0        2073     2     3     2   
363                1          0          0        2073     3     6     2   
319                4          0          0        1049     2     3     2   

           Cpu2  Cpu3  
laptop_ID              
750           9     9  
726          11    11  
633          11    11  
363           9     9  
319           8     8

In [101]:
X_pred.shape

(391, 17)

**¿Por qué da error?** 

In [104]:
predictions_submit = model.predict(X_pred)
predictions_submit

array([1256.12065574, 1219.68986423,  830.62298922, 1092.68524292,
        651.20426282,  628.27029985,  658.41874527, 1853.54106351,
       1932.43264401, 1199.18457019,  837.34816932, 1256.23212753,
       1219.72223614,  982.37207811,  727.72757412, 1045.61959255,
       1252.90061064, 2126.61646728,  664.31540417,  860.38598219,
       1091.62501974,  860.17774547,  599.14208062,  967.58402512,
        401.03731738, 1226.58829628,  621.06156747,  518.71348317,
        383.0332379 , 1199.18457019, 1057.11099674,  992.94729074,
       1014.37870895, 1047.3967399 , 1476.0939663 , 1291.82325195,
       1102.74520921, 1199.93830303,  582.72657392, 1944.30187958,
       1781.85276075, 1199.18457019,  534.80710486, 1356.17964483,
       1174.47532886, 2024.34493165, 1063.40960198, 1797.10683594,
       1087.24937946, 1789.45642829,  624.96690013,  331.24955409,
        854.26551514,  975.80465467,  457.26468781,  693.73814403,
        992.32121545,  647.42959792, 1161.08422375, 1286.54299

IMPORTANTE: APLICAR LO MISMO A ESTOS DATOS QUE HAYÁIS APLICADO A LOS DATOS DE ENTRENAMIENTO

- SI EL ARRAY CON EL QUE HICISTEIS `.fit()` ERA DE 4 COLUMNAS, PARA `.predict()` DEBEN SER LAS MISMAS
- SI AL ARRAY CON EL QUE HICISTEIS `.fit()` LO NORMALIZASTEIS, PARA `.predict()` DEBÉIS NORMALIZARLO
- TODO IGUAL SALVO BORRAR FILAS, EL NÚMERO DE ROWS SE DEBE MANTENER EN ESTE SET, PUES LA PREDICCIÓN DEBE TENER 391 FILAS, SI O SI

**Entonces, si al cargar los datos de train usé `index_col=0` para que utilizara la primera columna del conjunto de datos como índice, ¿tendré que hacerlo también para el conjunto `test.csv`?**

![](https://www.tierraljelechu.com/web/img/wiki_up/1.996-SorpresaDto.-1-Red.jpg)

In [105]:
X_pred = pd.read_csv("data/test2.csv", index_col=0)
X_pred.head()

Company  TypeName  Inches  Ram  OpSys  Weight  Memory1_1   
laptop_ID                                                             
750             10         5    14.0    8      6    1.25        256  \
726             15         5    14.0    4      6    1.40         32   
633             15         5    14.0    2      6    1.40         32   
363              2         2    15.6    8      6    2.65       1000   
319              5         1    11.6    2      6    1.10         32   

           Memory1_2  Memory2_1  Memory2_2  Resolution  Gpu1  Gpu2  Cpu1   
laptop_ID                                                                  
750                2          0          0        2073     2     3     2  \
726                1          0          0        2304     2     3     2   
633                4          0          0        2073     2     3     2   
363                1          0          0        2073     3     6     2   
319                4          0          0        1049     2     3     2   

           Cpu2  Cpu3  
laptop_ID              
750           9     9  
726          11    11  
633          11    11  
363           9     9  
319           8     8

### 3. AHORA puedo hacer la predicción que será lo que subirás a Kaggle. 

**¿Qué es lo que subirás a Kaggle?**

In [107]:
predictions_submit = model.predict(X_pred)
predictions_submit

array([1256.12065574, 1219.68986423,  830.62298922, 1092.68524292,
        651.20426282,  628.27029985,  658.41874527, 1853.54106351,
       1932.43264401, 1199.18457019,  837.34816932, 1256.23212753,
       1219.72223614,  982.37207811,  727.72757412, 1045.61959255,
       1252.90061064, 2126.61646728,  664.31540417,  860.38598219,
       1091.62501974,  860.17774547,  599.14208062,  967.58402512,
        401.03731738, 1226.58829628,  621.06156747,  518.71348317,
        383.0332379 , 1199.18457019, 1057.11099674,  992.94729074,
       1014.37870895, 1047.3967399 , 1476.0939663 , 1291.82325195,
       1102.74520921, 1199.93830303,  582.72657392, 1944.30187958,
       1781.85276075, 1199.18457019,  534.80710486, 1356.17964483,
       1174.47532886, 2024.34493165, 1063.40960198, 1797.10683594,
       1087.24937946, 1789.45642829,  624.96690013,  331.24955409,
        854.26551514,  975.80465467,  457.26468781,  693.73814403,
        992.32121545,  647.42959792, 1161.08422375, 1286.54299

**¡PERO! Para subir a Kaggle la predicción, ésta tendrá que tener una forma específica y no valdrá otra.**

En este caso, la **MISMA** forma que `sample_submission.csv`. 

In [108]:
sample = pd.read_csv("data/sample_submission.csv")

In [109]:
sample.head()

laptop_ID  Price_euros
0        750          500
1        726          500
2        633          500
3        363          650
4        319          650

In [22]:
sample.shape

(391, 2)

### 4. Mete tus predicciones en un dataframe. 

En este caso, la **MISMA** forma que `sample_submission.csv`. 

In [110]:
submission = pd.DataFrame({"laptop_ID": sample['laptop_ID'], "Price_euros": predictions_submit})

In [27]:
submission.head()

id  Price_euros
0  1184         2158
1   815          511
2  1244         5290
3  1121          622
4   443         1140

In [111]:
submission.shape

(391, 2)

### 5. Pásale el CHEQUEATOR para comprobar que efectivamente está listo para subir a Kaggle.

In [112]:
def chequeator(df_to_submit):
    """
    Esta función se asegura de que tu submission tenga la forma requerida por Kaggle.
    
    Si es así, se guardará el dataframe en un `csv` y estará listo para subir a Kaggle.
    
    Si no, LEE EL MENSAJE Y HAZLE CASO.
    
    Si aún no:
    - apaga tu ordenador, 
    - date una vuelta, 
    - enciendelo otra vez, 
    - abre este notebook y 
    - leelo todo de nuevo. 
    Todos nos merecemos una segunda oportunidad. También tú.
    """
    if df_to_submit.shape == sample.shape:
        if df_to_submit.columns.all() == sample.columns.all():
            if df_to_submit.laptop_ID.all() == sample.laptop_ID.all():
                print("You're ready to submit!")
                submission.to_csv("submission.csv", index = False) #muy importante el index = False
                urllib.request.urlretrieve("https://i.kym-cdn.com/photos/images/facebook/000/747/556/27a.jpg", "gfg.png")     
                img = Image.open("gfg.png")
                img.show()   
            else:
                print("Check the ids and try again")
        else:
            print("Check the names of the columns and try again")
    else:
        print("Check the number of rows and/or columns and try again")
        print("\nMensaje secreto de Clara: No me puedo creer que después de todo este notebook hayas hecho algún cambio en las filas de `diamonds_test.csv`. Lloro.")


In [113]:
chequeator(submission)

You're ready to submit!
